contrib/babelfishpg_tds/src/backend/tds/tdsbulkload.c

```diff
@@ -929,24 +1010,8 @@ ProcessBCPRequest(TDSRequest request)
 		}
 		PG_CATCH();
 		{
-			int			ret = 0;
-
-			HOLD_CANCEL_INTERRUPTS();
-
-			/*
-			 * Discard remaining TDS_BULK_LOAD packets only if End of Message
-			 * has not been reached for the current request. Otherwise we have
-			 * no TDS_BULK_LOAD packets left for the current request that need
-			 * to be discarded.
-			 */
-			if (!TdsGetRecvPacketEomStatus())
-				ret = TdsDiscardAllPendingBcpRequest();
-
-			RESUME_CANCEL_INTERRUPTS();
-
-			if (ret < 0)
-				TdsErrorContext->err_text = "EOF on TDS socket while fetching For Bulk Load Request";
-
+			CleanupBCPDuringError(internal_sp_started, before_subtxn_id,
```

Comment:
I think it's possible to combine the two TRY..CATCH block into a single one. The cleanup is working in the same way. 

```
while(1)
{
   PG_TRY();
   {
        message = SetBulkLoadRowData(req, message);

        /*
         * If the row-count is 0 then there are no rows left to be inserted.
         * We should begin with cleanup.
         */
        if (req->rowCount == 0)
        {
            /* Using Same callback function to do the clean-up. */
            pltsql_plugin_handler_ptr->bulk_load_callback(0, 0, NULL, NULL);
            break;
        }
    
        /* should be valid as rowCount > 0 */
        Assert(req->rowData);

        nargs = req->colCount * req->rowCount;
        values = palloc0(nargs * sizeof(Datum));
        nulls = palloc0(nargs * sizeof(bool));
    
        /* Flaten and create a 1-D array of Value & Datums */
        foreach(lc, req->rowData)
        {
            BulkLoadRowData *row = (BulkLoadRowData *) lfirst(lc);
    
            for (int currentColumn = 0; currentColumn < req->colCount; currentColumn++)
            {
                if (row->isNull[currentColumn]) /* null */
                    nulls[count] = row->isNull[currentColumn];
                else
                    values[count] = row->columnValues[currentColumn];
                count++;
            }
        }
    
        retValue += pltsql_plugin_handler_ptr->bulk_load_callback(req->colCount,
                                                                    req->rowCount, values, nulls);
    
        /* Free the List of Rows. */
        list_free_deep(req->rowData);
        req->rowData = NIL;

        if (values)
            pfree(values);
        if (nulls)
            pfree(nulls);
    }
    PG_CATCH();
    {
        int         ret = 0;

        /* hold interrupts during cleanup */
        HOLD_INTERRUPTS();
        CleanupBCPDuringError(..);
        RESUME_INTERRUPTS();

        PG_RE_THROW();
    }
    PG_END_TRY();
}
```

Reply:
I think there's some misunderstanding. I probably mentioned that context can be set beforehand.
```
volatile char *tmpContext = NULL;

tmpContext = "foo";
SetBulkLoadRowData();

tmpContext = "bar";
pltsql_plugin_handler_ptr->bulk_load_callback();

CATCH()
TdsErrorContext->err_text = tmpContext;
```